File: leetcode/0033.cpp
class Solution {
public:
    int search(vector<int> &nums, int target) {
        if (nums.size() == 0) return -1;
        auto low = nums.begin(), high = nums.end() - 1;
        while (low < high) {
            auto mid = low + (high - low) / 2;
            if (*mid >= *low) {
                if (*low <= target && target <= *mid)
                    high = mid;
                else
                    low = mid + 1;
            } else {
                if (*high >= target && target > *mid)
                    low = mid + 1;
                else
                    high = mid;
            }
        }
        if (*low == target) return low - nums.begin();
        return -1;
    }
};

File: leetcode/0033.go
package leetcode

func search(nums []int, target int) int {
	l, r := 0, len(nums)-1

	for l < r {
		mid := l + (r-l)/2
		if nums[mid] >= nums[l] {
			if target >= nums[l] && target <= nums[mid] {
				r = mid
			} else {
				l = mid + 1
			}
		} else {
			if target > nums[mid] && target <= nums[r] {
				l = mid + 1
			} else {
				r = mid
			}
		}
	}
	if nums[l] == target {
		return l
	}
	return -1
}
